src/modules/credential/cert/path_login.rs

```diff
@@ -0,0 +1,642 @@
+                if entry.certificate.is_empty() {
+                    log::error!("failed to parse certificate, name: {}", name);
+                    continue;
+                }
+
+                if entry.ocsp_enabled {
+                    ocsp_config.enable = true;
+                    ocsp_config.servers_override.extend(entry.ocsp_servers_override.iter().cloned());
+                    ocsp_config.failure_mode = if entry.ocsp_fail_open {
+                        ocsp::FailureMode::FailOpenTrue
+                    } else {
+                        ocsp::FailureMode::FailOpenFalse
+                    };
+                    ocsp_config.query_all_servers |= entry.ocsp_query_all_servers;
+                }
+
+                let mut certs = entry.certificate.clone();
+                certs.extend(entry.ocsp_ca_certificates.clone());
+
+                if is_ca_cert(&certs[0]) {
+                    for cert in certs.iter() {
+                        root_store_builder.add_cert(cert.clone())?;
+                    }
+                    trusted.push(ParsedCert { entry, certs });
+                } else {
+                    trusted_non_ca.push(ParsedCert { entry, certs });
+                }
+            }
+        }
+
+        Ok((root_store_builder.build(), trusted, trusted_non_ca, ocsp_config))
+    }
+
+    fn validate_cert(&self, roots: &X509Store, peer_certs: &[X509]) -> Result<Vec<X509>, RvError> {
+        if peer_certs.is_empty() {
+            return Ok(Vec::new());
+        }
+        let mut stack = Stack::<X509>::new()?;
+        peer_certs.iter().skip(1).try_for_each(|crt| stack.push(crt.clone()))?;
+
+        let mut context = X509StoreContext::new()?;
+        let (verified_res, verified_error, verified_chains) = context.init(roots, &peer_certs[0], &stack, |ctx| {
+            let ret = ctx.verify_cert()?;
+            let verified_chains: Vec<X509> = ctx.chain()
+                .map(|chain| chain.iter().map(|crt| crt.to_owned())
+                     .filter(|crt| !has_x509_ext_key_usage(crt) ||
+                             has_x509_ext_key_usage_flag(crt, XKU_ANYEKU | XKU_SSL_CLIENT))
+                     .collect())
+                .unwrap_or_else(Vec::new);
+
+            Ok((ret, ctx.error(), verified_chains))
+        })?;
+
+        if !verified_res {
+            return match verified_error.as_raw() {
+                X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT | X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY => {
+                    if peer_certs[0].not_after() <= Asn1Time::days_from_now(0)? {
+                        Err(rv_error_response!(& unsafe {X509VerifyResult::from_raw(X509_V_ERR_CERT_HAS_EXPIRED).to_string() }))
+                    } else {
+                        Ok(Vec::new())
+                    }
+                },
+                _ => Err(rv_error_response!(&verified_error.to_string())),
+            };
+        }
+
+        Ok(verified_chains)
+    }
+
+    fn matches_constraints(
```

Comment:
why there are so many constraints to be matched?

Reply:
> why there are so many constraints to be matched?

Vault can configure these matching items. In order to be compatible with Vault...